### ISSUE TITLE
What caret positions do we need to allow?

### ISSUE THREAD

**Issue by johanneswilm** [w3c/editing#51]
It has been requested that also caret movements are moved into intention events and not directly handled by contentEditableTyping. This makes sense.
Instead, I would like to put more emphasis on where the caret can actually go. It is currently quite annoying that there are many places where the caret cannot go or disappears and reappears. Does the following cover every place:

<p>All positions in which the caret can be placed programmatically.
The placement is restricted to be in one of these positions:</p>

<ol>
<li>Before or after any character in any text node.</li>

<li>After any inline element that is not a text node.</li>

<li>Before any inline element that is not a text node and
that does not have a `previousSibling`.</li>

<li>Inside any empty inline or block element that is not a
stub-element or inside any empty text node.</li>

<li>Before or after a block element that is also a stub
element.</li>

<li>In the <a>CE Typing Container</li> itself, if it is
empty.</li>
</ol>

<p>There are two exceptions to these rules:</p>

<ol>
<li>Within a table, the caret can only be placed inside
`TH` and `TD` elements.</li>

<li>The caret cannot be placed inside<del>, before or after a  `comment node` or a node that has it's display-value set to `none`.</del><ins> invisible nodes.</ins></li>

</ol>

**Comment by rniwa** [comment]
I'm not sure what you're referring to by "caret position". Are you talking about what JavaScript can set in getSelection().addRange?  Or you're talking about where the browser places caret when a user clicks somewhere moves selection via keyboard and other input devices?

**Comment by johanneswilm** [comment]
@rniwa If you try contenteditable right now, there are numerous places where the caret simply cannot go or at least isn't painted -- no matter whether put there programmatically or directly by movement. See for example this example: https://bug873883.bugzilla.mozilla.org/attachment.cgi?id=751510 . We also referred to this bug in this article: http://dl.acm.org/citation.cfm?id=2723154&CFID=675158185&CFTOKEN=70294709 . 

This is probably one of the top 5 bugs that make working with contenteditable near impossible in 2015. That's why we need an authoritative list of places where the caret needs to be able to be positioned for ContentEditableTyping.

**Comment by kojiishi** [comment]
Your item 6 of the list confused me...are we talking about within ce=typing, or are we talking about outside (ce=true) as well?

If the former, I think the basic idea is to give JS full control. Browsers raise events to ask JS to determine, and JS gives enough information back to browsers for it to render carets. When to raise what events, with what information, and what JS needs to give back, are what we need to discuss.

If the latter, I'm not very positive to solve it for now. It's so hard to make it interoperable at this point that we started ce=typing, hoping all editor developers move to ce=typing, and all authors move to editor frameworks based on ce=typing. If it succeeds, there should be no motivations to fix ce=true, right?

**Comment by johanneswilm** [comment]
@kojiishi 

> Your item 6 of the list confused me...are we talking about within ce=typing, or are we talking about outside (ce=true) as well?

This is all about ce=typing. Whether you want to fix ce=true as well is up to you. A corresponding bug has been filed to Chrome 2 years ago: https://code.google.com/p/chromium/issues/detail?id=242110

> If the former, I think the basic idea is to give JS full control.

Yes, JS has full access. But it still needs to be able to put the caret there. This is not about where the user automatically goes. It's a list of places where it needs to be able to go if put there pro grammatically. Right now there are many places where it cannot be put at all, so we need to avoid that happens again with ce=typing. That's why it needs to be in the spec.

**Comment by kojiishi** [comment]
Thanks for the clarification.

If we were talking about ce=typing, and when set from JS, IIUC, the request is browsers should never ever change the set selection, and it's up to JS (beforeinput) to make sure a text node is there. Correct?

If browsers do default cursor movements, I understand we need some logic defined. I didn't really carefully reviewed yet, but it looks like a good start. Are we also defining how to move when the caret fell into where not allowed?

**Comment by johanneswilm** [comment]
> If we were talking about ce=typing, and when set from JS, IIUC, the request is browsers should never ever change the set selection, and it's up to JS (beforeinput) to make sure a text node is there. Correct?

@kojiishi Correct. I don't know what goes on in a browser, but there seem to be both problems rendering the caret in certain places (it's there, but it isn't drawn) and there are places where it simply cannot be placed (different for every browser). This list is just about to ensure that it's both possible to put the caret there (by JS) and that it's actaully drawn/rendered when it's there.

> If browsers do default cursor movements, I understand we need some logic defined. I didn't really carefully reviewed yet, but it looks like a good start. Are we also defining how to move when the caret fell into where not allowed?

Indeed. I wrote some paragraphs on that, but removed them again when it sounded like you just wanted to do movement intentions. They are still in an old commit, I can get them back. They are not perfect, but they should be better than what we have now, and they shouldn't matter much anyway if the JS can put the caret in all the above-mentioned places programmatically.

**Comment by teleclimber** [comment]
> 1. After any inline element that is not a text node.

If we allow this caret location, and cE=typing inserts characters by default on typing that means we have to allow cE=typing to insert DOM text nodes in our content as needed. Is that correct?

Also:

> The caret cannot be placed inside, before or after a `comment node` or a node that has it's display-value set to `none`.

Why can't you place the caret before or after something that is invisible?

`abc<span style="display:none">ghost content</span>def`

With that rule it means I can't place the caret between `c` and `d`.

**Comment by johanneswilm** [comment]
> > 1. After any inline element that is not a text node.
> 
> If we allow this caret location, and cE=typing inserts characters by default on typing that means we have to allow cE=typing to insert DOM text nodes in our content as needed. Is that correct?

Correct. If we allow the insertion of characters, we need to create text nodes if they don't exist beforehand.

I had been thinking of the alternative first: only allow the caret in places where there already are text nodes. I think this should also work in cases where one is creating new content. 

The problem I see with that is what happens if the user saves a document and reopens it. I would guess that in most cases editors would save the DOM structure as HTML. So all empty text strings will be lost and have to be manually recreated when document is reloaded to make sure that the caret can go in these places.

And these empty text nodes will have to be insertted quite a lot of places. For example, in between elements if we want the caret to be able to go there:

```
<img ...>|<img ...>
```

| = caret

or 

```
<i>|</i>
```

Altogether I think it would make most sense if we don't let cE=typing create any new content by default and instead only create intentions (this is what the specification currently says). Also notice that the default caret movement behavior that was previously proposed, and the "legal caret positions" are not the same thing: The default movement would not move it into all the positions that the caret could be put in by means of legal caret positions.

> > The caret cannot be placed inside, before or after a `comment node` or a node that has it's display-value set to `none`.
> 
> Why can't you place the caret before or after something that is invisible?
> 
> `abc<span style="display:none">ghost content</span>def`
> 
> With that rule it means I can't place the caret between `c` and `d`.

Right, I guess the wording is not very good. I tried to make sure that the same position would not be mentioned twice (these positions are already mentioned as before the first or last characters in the sibling text nodes), but I understand that the current wording may also be interpreted as saying that the caret cannot be put in these positions at all. I will try to reword this to make sure this isn't misunderstood.

**Comment by ghost** [comment]
Shouldn’t it be stated that the caret cannot be placed inside invisible nodes, regardless of the nature of their invisibility? There’s much more than just comments and `display: none`: PI nodes (in XHTML), unparsed text (e. g. within `script`), nodes with `visibility: hidden`, nodes with `hidden` attribute set to true. Having the caret inside any of those doesn’t make much sense. Why the specificity?

**Comment by johanneswilm** [comment]
You are right, that would be a better way of stating it. What about mentioning the same position twice? Was I correct in assuming that this should be avoided? I am thinking of position descriptions such “before the first letter inside a text node” and “before a text node”. They should be the same thing and not two distinct caret positions.

**Comment by ghost** [comment]
Yes, I totally agree; if it’s unambiguously the same position, it should not have multiple representations.

**Comment by johanneswilm** [comment]
I have updated the spec to allow caret placement on either side of an invisible node and at the same time make the description so general that it covers all types of invisible nodes.

But now I think I can recall @fredck had some use cases for an invisible cE=True container. Would that use case still be of importance to cE=Typing?

**Comment by ghost** [comment]
Sorry if it’d been discussed before, but the more I read it, the more the terminology bothers me. HTML 5 does not describe ‘inline’ and ‘block’ elements (nor ‘stub’, for that matter), so it’s open to confusion: does ‘inline’ refer to what used to be inline in HTML 4 or an actual rendering as specified by CSS (which retains ‘inline/block’ dichotomy)? Hence I’ve got to ask: what’s wrong with HTML 5 terminology?

**Comment by johanneswilm** [comment]
I see. `inline` and `block` refer to the terminology as used for the CSS display-value: http://dev.w3.org/csswg/css-display/#the-display-properties

If 'inline' and 'block' are not used in current HTML terminology and only in CSS, why is it confusing what these could be about?

What is the correct terminology for `stub`-element? I'm fine with changing the names of all of these, as long as the functionality is preserved.

**Comment by johanneswilm** [comment]
With stub elements is meant basically SVG, img, canvas and likely some more I cannot think of. How about `br`? Can that be treated as any other stub element?

In the HTML spec I just find `void` elements for elements that have no end tag. But that doesn't then apply to SVGs, for example, right?

**Comment by fredck** [comment]
> But now I think I can recall @fredck had some use cases for an invisible cE=True container. Would that use case still be of importance to cE=Typing?

The only case that comes to my mind is the "paste bin" we use currently in CKEditor. It is hidden cE=true container where we move the user into on pasting to grab the pasted data (used in some browsers only). I don't think this requirement applies to cE=typing.

**Comment by fredck** [comment]
> only allow the caret in places where there already are text nodes

This would be extremely complex to handle (as you also stated in [your comment](https://github.com/w3c/editing/issues/51#issuecomment-111752249)).

The caret is simply a UI representation of the selection. It should be possible to see it in all contexts where we expect users to be able to "type". This doesn't include invisible elements simply because we don't expect users to type then, but if we don't make it a restriction it should not be a problem anyway. I mean, if one would set the selection inside an invisible element, then typing would still happen there. The thing that would block this situation to happen is not really cE=typing, but the behavior of the caret to input commands, like arrow-keys and mouse click, which would never make a selection inside invisible elements. So sincerely, I would not care about blocking typing inside invisible elements. For the scope of cE=typing, I would just think about the selection position. The problem of placing the caret in allowed positions should be instead handled by the Selection API.

Still it should be possible to place the caret at any position where text can be inserted. cE=typing is all about this and if a text node is not available for that position, it should be created on the fly.

If cE=typing will not create any node by default, then it is senseless for it to exist as an editing module. We should then review the whole thing and introduce cE=intent, where UAs would not have any default behavior and just intentions would be fired.

**Comment by johanneswilm** [comment]
> The problem of placing the caret in allowed positions should be instead handled by the Selection API.

We can move this part to another spec, but we still need to have this down precisely. Right now, as you know, the caret cannot be placed in certain places, so we need certain "guarantees" that the caret really can be placed in specific places with a new standard. Of course, the caret is a visual manifestation of the currently collapsed selection, so that would be an argument for moving it there, but then again, that would affect all of cE=True as well which may not be wanted. 

> If cE=typing will not create any node by default, then it is senseless for it to exist as an editing module. We should then review the whole thing and introduce cE=intent, where UAs would not have any default behavior and just intentions would be fired.

Right now that is kind of what cE=typing is. But we can change the name of it. I am not sure about the rest of this paragraph: are you saying we need to have something else in terms of functionality?

**Comment by rniwa** [comment]
I agree this should be spec'ed in selection API.

**Comment by ghost** [comment]
Some of those limitations (e. g. no caret between block elements) may harm Selection’s flexibility in general. How are they supposed to be spec’ed in the Selection API?

**Comment by ghost** [comment]
As for ‘inline’ and ‘block’, CSS terminology may imply it’s CSS-dependent; from the Display Module perspective, a `<p style="display: inline">` is ‘inline’, not ‘block’ as we’d expect in HTML Editing APIs/HTML 4 terms. That’s sort of ambiguous.

Per HTML 5, most illegal positions may be described as follows: Elements that do not accept text/flow content must not accept the caret (in addition to `table`, those are `ul`, `ol` and `dl`). In ‘inline/block’ dichotomy, we’ve got to describe those as exceptions, which is a pain (we missed a legal `caption` within tables) and less future-proof (while UAs are already on their way to HTML 5.1).

‘Stubs’ are not voids, right. They just don’t accept any content (`br` fits here). Note: SVG (and MathML, for that matter) is not stub: it may hypothetically have editable text nodes down the tree (that’s complicated, maybe not for v1).

Text nodes are also not as simple as they look. HTML defines an ‘inter-element whitespace’ (represented in DOM as text node, but not considered text by HTML); they are allowed almost anywhere, but will immediately invalidate the document on any non-whitespace input when in certain positions (e. g. direct child of a `table`). They fit the definition of ‘any text node’ [1, 4], but they shouldn’t be a legal caret position.

**Comment by johanneswilm** [comment]
> As for ‘inline’ and ‘block’, CSS terminology may imply it’s CSS-dependent; from the Display Module perspective, a `<p style="display: inline">` is ‘inline’, not ‘block’ as we’d expect in HTML Editing APIs/HTML 4 terms. That’s sort of ambiguous.

Well, yes. The way it is defined now, it is dependent on CSS. Things that are invisible with `display:none` cannot contain a caret, those turned into inline elements will behave as such and so on. On my part, this was on purpose. 

> Per HTML 5, most illegal positions may be described as follows: Elements that do not accept text/flow content must not accept the caret (in addition to `table`, those are `ul`, `ol` and `dl`). In ‘inline/block’ dichotomy, we’ve got to describe those as exceptions, which is a pain (we missed a legal `caption` within tables) and less future-proof (while UAs are already on their way to HTML 5.1).

If you think it's easier to describe caret positions in terms of what is illegal, I'm OK with. The important thing is that it can go everywhere where they cannot always go today (for example, in-between SVGs or canvas-elements). Could you come up with an entire text about legal caret positions that could replace the current text and at the same time incorporate all the positions that are currently defined as legal?

> Some of those limitations (e. g. no caret between block elements) may harm Selection’s flexibility in general. 

To me at least this limitation doesn't _have_ to be there. The current list is just the minimum amount of places where the caret has to be able to go. This is important to avoid the current situation in which the caret can go in different places in each browser. All these are only places where the caret can be placed programmatically, so it's fully up to the JS code whether it wants to make use of a particular place or not

> Text nodes are also not as simple as they look. HTML defines an ‘inter-element whitespace’ (represented in DOM as text node, but not considered text by HTML); they are allowed almost anywhere, but will immediately invalidate the document on any non-whitespace input when in certain positions (e. g. direct child of a `table`). They fit the definition of ‘any text node’ [1, 4], but they shouldn’t be a legal caret position.

Ok, is that the only exception for text nodes? Is there a name for the other text nodes?

> Note: SVG (and MathML, for that matter) is not stub: it may hypothetically have editable text nodes down the tree (that’s complicated, maybe not for v1).

Indeed, but SVGS have a different internal namespace. For now I would like to ensure that the caret can be placed on either side of an SVG under all circumstances (currently it is not possible to put the caret inbetween two inline SVGs in some browsers). More advanced things can wait for future versions.

**Comment by rniwa** [comment]
I don't think we can just make up rules. We need to study what existing browsers allow, and document that with some fixes.

**Comment by johanneswilm** [comment]
@rniwa I am not sure exactly what you want to document, but do you want to do that documentation? I have started some with the bug reports I filed here: https://bugzilla.mozilla.org/show_bug.cgi?id=873883 .

**Comment by ghost** [comment]
> The way it is defined now, it is dependent on CSS. [...] On my part, this was on purpose.

This makes sense to me. But it’s also not impossible to think those are HTML 4/Historical Editing APIs terms, no?

> If you think it's easier to describe caret positions in terms of what is illegal, I'm OK with.

No, it’s a bit different. Some positions will depend on context (CSS-dependent ones, plus what you said about ‘the same position twice’), and some are _always_ illegal (those are currently listed as exceptions). I just don’t want those element names be spec’ed explicitly, as that’s a fragile way. The idea is that they cannot accept any text per HTML, therefore they cannot accept the caret per cE=typing, _regardless of their actual names_ (that includes `table`, `ul`, `ol`, `dl`, all stubs and whatever we may get in HTML 5.11). Unfortunately, I cannot replace the whole text with a simple line; it’s just that exceptions may be shorter. Do you see what I mean?

> To me at least this limitation doesn't _have_ to be there.

In Selection API or in implementations? I mean, Selection API allows nearly anything, it doesn’t give a damn about the caret or editing, and that’s how it should be. For editing, browsers impose restrictions, and so do we, but those are practical limitations, not abstract ones. That’s why I believe _we_, not Selection API, should spec them. Where am I wrong?

> Is that the only exception for text nodes?

One more that comes to mind is what happens to a collapsed whitespace. As multiple whitespace characters render as a single space by default, they should be treated as such by the caret (that’s how it works now, but that’s not spec’ed, again); but if they are visible (e. g. `white-space: pre-wrap`), they are treated as multiple spaces, i. e. the caret moves between them.

> Is there a name for the other text nodes?

Per HTML, the difference between inter-element whitespace and normal text is that normal text is in the ‘Palpable content’ category, while inter-element whitespace is not (http://www.w3.org/TR/html5/single-page.html#palpable-content). Is that what you mean?

> We need to study what existing browsers allow, and document that with some fixes.

Why? What browsers allow varies between vendors, it’s based exactly on made up rules, for each their own. I mean, Gecko allows more positions than WebKit does, and some of those positions are pretty meaningless from the editing point of view; why document them?

**Comment by johanneswilm** [comment]
> > The way it is defined now, it is dependent on CSS. [...] On my part, this was on purpose.
> 
> This makes sense to me. But it’s also not impossible to think those are HTML 4/Historical Editing APIs terms, no?

Yes, I can see how this could lead to confusion. If we are going to continue down this current path, we should probably be more specific that these values come from CSS somehow.

> > To me at least this limitation doesn't _have_ to be there.
> 
> In Selection API or in implementations? I mean, Selection API allows nearly anything, it doesn’t give a damn about the caret or editing, and that’s how it should be. For editing, browsers impose restrictions, and so do we, but those are practical limitations, not abstract ones. That’s why I believe _we_, not Selection API, should spec them. Where am I wrong?

To take a concrete example: caret placement between block elements. I cannot see how editors need them, but if browser makers insist on adding these in the spec and in implementations so that carets are more similar to selections in general, then I don't see how they do much harm either, as the JS code simply can be made to not take advantage of those extra possible placements.

I agree that this is currently how the separation is. All I am saying is that if there strong will to extend the Selection API to also treat carets, then I am not strongly opposed to it, as long as it is still specified somewhere. For me it is most important that it moves forward in some way.

That being said, I can also see some good reasons why _we_ should continue to have it here. For example, it was earlier said that we cannot fix cE=true because too much legacy JS code depends on the current, broken behavior.   But if we add it to the selection API, then it will apply to all editing context, which means it will also change how cE=true works and thereby likely break this legacy code.

> One more that comes to mind is what happens to a collapsed whitespace. As multiple whitespace characters render as a single space by default, they should be treated as such by the caret (that’s how it works now, but that’s not spec’ed, again); but if they are visible (e. g. `white-space: pre-wrap`), they are treated as multiple spaces, i. e. the caret moves between them.

True. At the same time zero-width spaces are treated as individual characters. But this is more about the part of caret movement, right? It's not a problem if the caret can be programmatically placed anywhere inside a long series of white spaces, is it?

So if we add caret movement directly to this spec, then this should be taken into consideration (and I think the "move by at least 1 px" rule that was in the prveious draft would actually do that).

**Comment by ghost** [comment]
> We should probably be more specific that these values come from CSS somehow.

Yes, that’s what I mean. It’s just a matter of clear definitions.

>  If there strong will to extend the Selection API to also treat carets, then I am not strongly opposed to it.

I’m not strongly opposed either, I just think we are more specific here than Selection should be; some of those concepts may map onto Selection pretty well, but some may not. Now, if everybody else see that differently, who am I to oppose? As long as it’s spec’ed somewhere, I’m fine. But if we just mail that to Selection API, we may never hear back, and it will remain unspec’ed; that’s something I’d like to avoid.

> This is more about the part of caret movement, right? It's not a problem if the caret can be programmatically placed anywhere inside a long series of white spaces, is it?

I think so. The only problem is that on the caret between some 8 spaces, there will be spaces popping out of nowhere on input (as a result of de-collapsing). Not really the best experience for editors.
